src/connectors/assetpanda.py

```diff
@@ -0,0 +1,173 @@
+    for key, value in replace_key.items():
+        for device in list_device:
+            if device.get(key, False):
+                if device.get(value) is not None:
+                    number = 2
+                    while device.get(f"{value}_{number}") is not None:
+                        number += 1
+                    value = f"{value}_{number}"
+                device[value] = device.pop(key)
+    return list_device
+
+
+### Main Functions ###
+
+def get_data(token: str, url: str, params: dict = {}) -> dict:
+    headers: dict = {
+        "Authorization": f"Bearer {token}"
+    }
+    try:
+        log.debug(f"Preparing GET: url={url} with params={params}")
+        req = requests.get(url, params=params, headers=headers)
+        print("req: ", req)
+        req.raise_for_status()
+    except HTTPError as http_err:
+        log.error(f"Error GET: url={url}")
+        log.error(f"HTTP error occurred: {http_err}")
+        raise http_err
+    log.debug(req.status_code)
+
+    return req.json()
+    
+    
+def connect(connection_name, options):
+    landing_table = f'data.assetpanda_{connection_name}_connection ' # creates table in snowalert
+    
+    comment = yaml_dump(module='assetpanda', **options)
+
+    db.create_table(name=landing_table, cols=LANDING_TABLE_COLUMNS, comment=comment)
+
+    db.execute(f'GRANT INSERT, SELECT ON {landing_table} TO ROLE {SA_ROLE}')
+
+    return {
+        'newStage': 'finalized',
+        'newMessage': "AssetPanda ingestion tables created!",
+    }
+
+
+def ingest(table_name, options):
+    landing_table = f'data.{table_name}'
+    
+    token = options['token']
+    asset_entity_id = options['asset_entity_id']
+
+    general_url = f"https://api.assetpanda.com:443//v2/entities/{asset_entity_id}/objects"
+    fields_url  = f"https://api.assetpanda.com:443//v2/entities/{asset_entity_id}"
+
+    params = {
+        "offset": 0,
+        "limit": PAGE_SIZE,
+    }
+
+    total_object_count = 0 
+
+    while params['offset'] <= total_object_count:
+
+        log.debug("total_object_count: ", total_object_count)
+
+        assets = get_data(token=token, url= general_url, params=params)
+        
+        list_object, total_object_count = get_list_objects_and_total_from_get_object(assets)
+
+        dict_fields = get_data(token, fields_url, params=params)
+        list_field = dict_fields["fields"]
+
+
+        # Stripping down the metadata to remove unnecessary fields. We only really care about the following:
+        # {"field_140": "MAC_Address", "field_135" :"IP"}
+        clear_fields: list = reduce(reduce_fields, list_field, {})        
+
+        # replace every key "field_NO" by the value of the clear_field["field_NO"]
+        list_object_without_field_id = replace_device_key(list_object, clear_fields)
+
+
+        db.insert(
+            landing_table,
+            values=[(
+                entry,
+                entry.get('id', None),
+                datetime.utcnow()
```

Comment:
This is going to give every page you insert a slightly different insert time. This will make it a bit challenging to get all of the latest results; you have to select where the insert_time is within a certain range and make sure your range doesn't include any results from the previous scan. If you mark an insert_time before you get into the `while` loop, you can use that insert_time for each page and then select from the table where insert_time = max(insert_time).